tmt/base.py

```diff
@@ -978,7 +978,13 @@ def __init__(
 
         # Store 'environment' and 'environment-file' keys content
         self._environment = self._get_environment_vars(node)
-        # Expand all environment variables in the node
+        # Store context variables in a dictionary se we don't have to
+        # build it over and over again
+        self._dict_context_var = {
```

Comment:
1. would it be possible to not store the mapping in self, but rather pass it to method as its second argument? It seem its only purpose is to be consumed by the method, called right after it, never to be used again, and having it as a parameter would remove one shared mutable state, attached to the instance, and the method would be slightly purer.
2. 2. (nitpicking…) WRT the name - a type in the name (“dict”) is very uncommon in python lands, “context” - ther’s so many context, “fmf_ context” would be more fitting.

And by solving 1. we could solve 2. easily, prepare your mapping and pass it tomethod as a “fmf_context” parameter :) No shared state, purer methods, making them easier to test, no loose ends and it’s immediatel clear which context we deal with here :)

Reply:
@happz Updated. I am not sure about type annotations though.

Reply:
Thank you, it definitely looks better. `Dict[str, str]` is perfectly fine, hard to find a more fitting type annotation. Just two more comments:

* how about making the parameter not optional? AFAICT, seems to be part of each and every call to `_expand_node_data()`, therefore it's not really optional. Giving it a default value does not simplify any call, `fmf_context: Dict[str, str]` would be perfect.
* Not related to the function itself, but one Python trap: never use mutable defaults, like `foo: Dict[str, str] = {}`, as it leads to unexpected behavior. Something to be aware of for the future, more details in https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/

Reply:
OK. Updated in the latest commit. Thank you for your feedback.

Reply:
Thanks, LGTM!